qemu/tests/pci_bridge_hotplug_unplug.py

```diff
@@ -0,0 +1,142 @@
+
+
+def check_all_hotplugged_disks(test, params, env, image_num):
+    """
+    Check all hotplugged disks.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    error_context.context("Check 'info block'!", logging.info)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    monitor_info_block = vm.monitor.info_block(False)
+    if image_num != len(monitor_info_block.keys()):
+        test.fail("The amount of block devices in 'info block' is incorrect!")
+    logging.info("Checking 'info block' pass!")
+
+    error_context.context("Read and write data on all hotplugged disks!", logging.info)
+    sub_test_type = params.get("sub_test_type", "dd_test")
+    images = params.get("images")
+    images = images.split()
+    images.pop(0)
+    for image in images:
+        params["dd_if"] = "ZERO"
+        params["dd_of"] = image
+        utils_test.run_virt_sub_test(test, params, env, sub_test_type)
+
+        params["dd_if"] = image
+        params["dd_of"] = "NULL"
+        utils_test.run_virt_sub_test(test, params, env, sub_test_type)
+    logging.info("Read and write data on all hotplugged disks successfully!")
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    [pci-bridge] Hotplug and unplug different devices to 1 pci-bridge
+    [pci-bridge] Hotplug and unplug different devices to different pci-bridges
+    This case will:
+    1) Start the guest.
+    2) Hotplug different devices to pci-bridge(s).
+    3) Check devices.
+    4) Hot-unplug devices.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    pci_bridge_num = int(params.get("pci_bridge_num", 0))
+    if pci_bridge_num > 1:
+        raise NotImplementedError()
+
+    error_context.context("Get the main VM", logging.info)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    login_timeout = int(params.get("login_timeout", 360))
+    session = vm.wait_for_login(timeout=login_timeout)
+
+    image_list = params.get("images").split()
+    image_num = len(image_list)
+    if (image_num < 2):
+        test.fail("No pre-prepared images for hotplugging!")
+
+    device_list = []
+    fmt_list = params["disk_driver"].split()
+    if pci_bridge_num < 1:
+        test.fail("There should be at least one pci-bridge!")
+    elif pci_bridge_num == 1:
+        pci_bridge_id = params.get("pci_controllers").strip()
+
+        for i in range(image_num - 1):
+            if i >= len(fmt_list):
+                test.fail("No sufficient disk driver type!")
+            error_context.context("Hotplug a %s disk!" % fmt_list[i], logging.info)
+            device_list += disk_hotplug(test, vm, params, image_list.pop(), fmt_list[i], pci_bridge_id)
+            logging.info("Hotplug a %s disk successfully!" % fmt_list[i])
+        time.sleep(60)
```

Comment:
Occasionally， It need a time slot to be detected for nec-usb-xhci and usb-storage devices after hot-plugging.